app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb

```diff
@@ -350,13 +357,54 @@ def populate_hardware_hash_with_series_attributes(hardware_hash, instance, serie
         hardware_hash[:memory_mb]       = series[:memory] / 1.megabyte
         hardware_hash[:disk_capacity]   = series[:root_disk_size] + series[:swap_disk_size]
 
-        os_disk = instance.properties.storage_profile.os_disk
-        sz      = series[:root_disk_size]
-        vhd_loc = os_disk.try(:vhd).try(:uri) || os_disk.try(:managed_disk).try(:id)
+        add_azure_instance_disk(hardware_hash[:disks], instance)
+
+        # No data available on swap disk? Called temp or resource disk.
+      end
+
+      def add_azure_instance_disk(array, instance, os_disk = nil)
+        os_disk ||= instance.properties.storage_profile.os_disk
 
-        add_instance_disk(hardware_hash[:disks], sz, os_disk.name, vhd_loc) unless sz.zero?
+        if instance.managed_disk?
+          disk_type     = 'managed'
+          disk_location = os_disk.managed_disk.id
+          managed_disk  = @managed_disks.find { |d| d.id.casecmp(disk_location).zero? }
+          disk_size     = managed_disk.properties.disk_size_gb.gigabytes
+          mode          = managed_disk.sku.name
+        else
+          disk_type     = 'unmanaged'
+          disk_location = os_disk.try(:vhd).try(:uri)
+          disk_size     = os_disk.try(:disk_size_gb).try(:gigabytes)
+
+          if disk_location
+            uri = Addressable::URI.parse(disk_location)
+            storage_name = uri.host.split('.').first
+            container_name = File.dirname(uri.path)
+            blob_name = uri.basename
+
+            storage_acct = @storage_accounts.find { |s| s.name.casecmp(storage_name).zero? }
+            mode = storage_acct.sku.name
+
+            if disk_size.nil?
+              storage_keys = @sas.list_account_keys(storage_acct.name, storage_acct.resource_group)
+              storage_key  = storage_keys['key1'] || storage_keys['key2']
+              blob_props   = storage_acct.blob_properties(container_name, blob_name, storage_key)
+              disk_size    = blob_props.content_length.to_i
+            end
```

Comment:
why would disk size be nil? I would love to avoid the complexities around lines 389 - 392

Reply:
I haven't completely nailed down why some vm's include the disk size in their storage profile, and why some don't, but it must have something to do with the way they were provisioned, or possibly even *when* they were provisioned. Or it could have something to do with the OS. In any case, I cannot predict it at this time.

You can see the difference by running the following commands on the CLI:
```
az vm show -n miq-test-win12 -g miq-azure-test1 # This doesn't show a disk size.
az vm show -n dberger1 -g dberger1 # This does show a disk size
```